### ISSUE TITLE
Support baseUrl/path import aliases (and other newish ts/ts-node features)

### ISSUE THREAD

**Issue by tenwit** [pulumi/pulumi#8787]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Some of my typescript Pulumi codebases are monorepos, and suffer from relative path hell. Typescript provides the [path mapping sugar](https://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping) to help with this.

I'd also like to use ESM throughout my codebases. ts-node supports this, though currently only via [experimental settings](https://github.com/TypeStrong/ts-node/issues/1007).

Recent changes in Pulumi seem to allow supporting these features much simpler (e.g. https://github.com/pulumi/templates/issues/244, https://github.com/pulumi/pulumi/pull/8655), though it might not be possible yet (https://github.com/pulumi/pulumi/issues/4876).

Requests:
- Can new ts/ts-node features be turned on through suitable configuration of Pulumi.yaml (nodeargs) and tsconfig.json?
- Can docs describing how to do this be added to https://www.pulumi.com/docs/intro/languages/javascript/#typescript?

### Affected area/feature

Typescript support.

**Comment by tenwit** [comment]
Overlaps with #3061

**Comment by RobbieMcKinstry** [comment]
Hello @tenwit ! Coming across this issue just now, and I wanted to give a little more context into how I'm currently thinking about this issue and its blockers.

I'm reading this issue as having two main feature requests (plus documentation):

1. Can users pick a better version of TS-Node?
2. Can Pulumi improve support for monorepos by allowing them to use one `tsconfig` file at the project root?
3. More generally, can users have greater control over how the runtime executes their programs (via https://github.com/pulumi/pulumi/issues/3061)?

**TL;DR:**  IMO these three items are separate requests. I also think these items are tracked in other open issues. This is really exciting for me, because it indicates that these are features people _really_ want, if users are independently arriving at the same conclusions. I'm going to type up an explanation of the state of the world for the features as described above.

# More Flexible Dependencies

> Can new ts/ts-node features be turned on through suitable configuration of Pulumi.yaml 

This one has been on my radar for six months now, and I really want to improve things here. 🤗  The short answer is "no, that's currently a breaking change, but I wish it weren't so!". The long answer is under discussion [here](https://github.com/pulumi/pulumi/pull/10122). I'm hopeful we can come up with a slick solution. For example:

1. Import the current version of ts-node under an alias.
2. Use a dynamic import to see if the user has ts-node installed as a peer dependency. If so, use the version provided by the user.
3. If not, fallback to the aliased version.

Another approach that's rattled around in my mind, but I don't have a good design for, is some kind of "lifecycle hooks" for the Node runtime. A lot of the motivation for upgrading TS-Node and TypeScript versions is using one's own typechecker or transpiler for faster builds. I'd like to consider adding a configuration option to the TypeScript runtime that allows users to specify their own build step. For example, putting this in your `Pulumi.yaml` file:

```yaml
runtime:
  name: nodejs
  options:
    typescript: true
    buildStep: build
```

...where `buildStep: myCustomBuildStep` means "look in my `package.json` file for a script named `myCustomBuildStep` and use that instead of TS-Node."

# Better Monorepo Support

> Typescript provides the [path mapping sugar](https://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping) to help with this. 

RE: path mapping, I recently used path mapping in our generated Node code, but I came across [this open bug](https://github.com/Microsoft/TypeScript/issues/25677) in the TypeScript compiler which makes path mapping unusable in libraries. I suppose that as long as your Pulumi program as `tsconfig` available during runtime, you should still be able to use path mapping, right? I think you should be able to do this right now, though perhaps not in a monorepo with only one `tsconfig` file above the Pulumi project in the directory hierarchy. However, we're actively working on https://github.com/pulumi/pulumi/issues/2619 which will fix this.

I expect once https://github.com/pulumi/pulumi/issues/2619 is closed, you'll be able to use path mapping without issue. Does that sound right to you?

# Pass Args to Runtimes

> Overlaps with https://github.com/pulumi/pulumi/issues/3061

Right now, the Pulumi engine executes a Go binary called the "language host", which is responsible for spinning up each "language runtime" and providing the runtime with the appropriate flags. The good news is that way flags are passed between the `engine => lang host` and `lang host => lang runtime` is pretty well defined, so I don't think this issue has any obvious blockers. I think it'd be relatively straightforward to implement, but it requires a degree of design work, since it's not obvious to me what the CLI UX would look like for passing those flags. My point of reference is Cargo, which does something like this:

 ```bash
$ cargo run -- custom args --and -subcommands=true
```

The bad news is that this requires plumbing into all six of our language runtimes: NodeJS, Go, Python, Dotnet, YAML, and Java. That makes a little daunting of a task for one person, but because it's parallelizable and support for each language can be implemented in isolation of the others, maybe it's something we can collaborate on with a community contributor.

----------

If I understand your ask correctly, this issue is tracked in https://github.com/pulumi/pulumi/pull/10122, https://github.com/pulumi/pulumi/issues/2619, and https://github.com/pulumi/pulumi/issues/3061. Would it be fair to close this issue as a duplicate of those? 

I'm happy to continue the conversation here or in those issues. I'm actively working on https://github.com/pulumi/pulumi/issues/2619 , passively working on https://github.com/pulumi/pulumi/pull/10122, and unfortunately https://github.com/pulumi/pulumi/issues/3061 is backlogged for now.

**Comment by omercnet** [comment]
would love to see support for paths
it works when using pulumi automations, but not when running the project directly with pulumi

**Comment by cnunciato** [comment]
Removed the `awaiting-feedback` label as we probably don't need to wait for feedback before at least investigating this issue. (I just hit it myself looking into https://github.com/pulumi/pulumi-hugo/issues/1504.) 

>I expect once https://github.com/pulumi/pulumi/issues/2619 is closed, you'll be able to use path mapping without issue. 

#2619 has been closed, but this is indeed still an issue. Using `paths` does not work when running either `pulumi up` or with a local program using `stack.up()` with Automation API. (I still get `MODULE_NOT_FOUND`, despite the TypeScript compiler being happy.

```javascript
{
    "compilerOptions": {
        "strict": true,
        "outDir": "bin",
        "target": "es2016",
        "module": "commonjs",
        "moduleResolution": "node",
        "sourceMap": true,
        "experimentalDecorators": true,
        "pretty": true,
        "noFallthroughCasesInSwitch": true,
        "noImplicitReturns": true,
        "forceConsistentCasingInFileNames": true,
        "baseUrl": "../",
        "paths": {
            "myapp": [
                "app"
            ]
        }
    },
    "files": [
        "index.ts"
    ]
}
```

```
> start
> ts-node index.ts

Error: Cannot find module 'myapp'
Require stack:
- /Users/cnunciato/Projects/dev/pulumi/aws-typescript-38c78f1/runner/index.ts
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
    at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (/Users/cnunciato/Projects/dev/pulumi/aws-typescript-38c78f1/runner/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)
    at Function.Module._load (node:internal/modules/cjs/loader:901:27)
    at Module.require (node:internal/modules/cjs/loader:1115:19)
    at require (node:internal/modules/helpers:130:18)
    at Object.<anonymous> (/Users/cnunciato/Projects/dev/pulumi/aws-typescript-38c78f1/runner/index.ts:4:1)
    at Module._compile (node:internal/modules/cjs/loader:1241:14)
    at Module.m._compile (/Users/cnunciato/Projects/dev/pulumi/aws-typescript-38c78f1/runner/node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1295:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/cnunciato/Projects/dev/pulumi/aws-typescript-38c78f1/runner/node_modules/ts-node/src/index.ts:1621:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/cnunciato/Projects/dev/pulumi/aws-typescript-38c78f1/runner/index.ts'
  ]
}
```

**Comment by lesad** [comment]
Struggling to get the proposed solution to work. @cnunciato could you kindly elaborate on your example? 

I've tried to use `tsconfig-paths` but I still get the same errors. Could you provide the code-structure you used?

**Comment by cenk1cenk2** [comment]
I do this naive approach expandinf this.

Where the following function is exported from a shared library to be reusable.

```typescript
export async function entrypointWithTsConfigPaths(options: EntrypointWithTsConfigPathsOptions): Promise<unknown> {
  options = {
    tsconfig: '../',
    ...options
  }

  const paths = await import('tsconfig-paths')

  const config = paths.loadConfig(options.tsconfig)

  if (config.resultType === 'success') {
    paths.register({
      baseUrl: config.absoluteBaseUrl,
      paths: config.paths
    })
  } else {
    throw new Error('Failed to load tsconfig paths.')
  }

  return options.cb()
}
```

Then on the library I create a seperate file called `entrypoint.ts` and just call it.

```typescript
export = entrypointWithTsConfigPaths({
  cb: () =>
    import('./main').then((main) => {
      return main.default()
    })
})
```

Where in the `main.ts` I have my actual Pulumi endpoint.

This is then used by Pulumi through the `package.json` file main property.

```
"main": "./src/entrypoint.ts
```

This is mainly just for getting rid of it easily if this ever gets implemented in, so I did not include it in any Pulumi related code and keep it seperate.

Meanwhile hopefully it would be useful to you.

**Comment by saiintbrisson** [comment]
Would bumping the TypeScript version be acceptable in order to close this? It's the missing piece of the puzzle. Importing custom a ts-node `require` from the tsproject file is easy enough, the problem is that only "recent" `ts-node` versions accept setting custom requirements through code. And bumping ts-node requires bumping typescript.

**Comment by ps-mattstuart** [comment]
I'm running typescript 5 and ts-node 10.9.2 and it's definitely not fixed easily. I get the error:

```
    error: Error: Unable to find test
        at resolveConfigPath (.../infra-pulumi/node_modules/tsconfig-paths/src/tsconfig-loader.ts:88:4)
        at loadSyncDefault (.../infra-pulumi/node_modules/tsconfig-paths/src/tsconfig-loader.ts:57:22)
        at tsConfigLoader (.../infra-pulumi/node_modules/tsconfig-paths/src/tsconfig-loader.ts:46:22)
        at configLoader (.../infra-pulumi/node_modules/tsconfig-paths/src/config-loader.ts:66:22)
        at Object.register (.../infra-pulumi/node_modules/tsconfig-paths/src/register.ts:86:36)
        at Object.<anonymous> (.../infra-pulumi/node_modules/tsconfig-paths/register.js:1:15)
        at Module._compile (node:internal/modules/cjs/loader:1550:14)
        at node:internal/modules/cjs/loader:1702:10
        at Object.require.extensions.<computed> [as .js] (.../infra-pulumi/node_modules/ts-node/src/index.ts:1608:43)
        at Module.load (node:internal/modules/cjs/loader:1307:32)

```
Where `test` above is my project name and current path.
My tsconfig matches what's been described here, package.json and pulumi.yaml are all very straight forward. What DOES fix this problem is running Pulumi like this:

```
TS_NODE_PROJECT=$(pwd)/tsconfig.json pulumi preview`
```
And this is because of the following section in tsconfig-paths:
```
export function register(params?: RegisterParams): () => void {
  ...
  } else {
    // eslint-disable-next-line
    const minimist = require("minimist");
    const argv = minimist(process.argv.slice(2), {
      // eslint-disable-next-line id-denylist
      string: ["project"],
      alias: {
        project: ["P"],
      },
    });
    cwd = argv.project;
  }
  ...
```
The issue is that `--project` is being set by the pulumi binaries to whatever my current pulumi project is, `test` in this case, which is interfering with TSCONFIG-PATHS loading the correct tsconfig.json file.

The ultimate argv set that tsconfig-paths sees:

```
[
      '/opt/homebrew/Cellar/node/23.2.0/bin/node',
      '.../infra-pulumi/node_modules/@pulumi/pulumi/cmd/run/index.js',
      '--monitor',
      '127.0.0.1:61590',
      '--engine',
      '127.0.0.1:61583',
      '--sync',
      '/var/folders/q9/8zgzpbq90vgcfnpmrg8q66fr0000gn/T/pulumi-node-pipes580091113',
      '--organization',
      'pulumi-org',
      '--project',
      'test',
      '--stack',
      'dev',
      '--pwd',
      '.../infra-pulumi/test',
      '--dry-run',
      '--query-mode',
      'false',
      '--parallel',
      '96',
      '.'
    ]
```

**Comment by ps-mattstuart** [comment]
Part 2: I've switched to using the in-code version to get around this. I'm not the best at Typescript or node unfortunately as I was trying to replicate the tsconfig-paths/register functionality being auto loaded using ts-node:require but I can't get there with the time I have.

Ultimately I added

bootstrap.ts
```
import * as tsConfigPaths from "tsconfig-paths";

export function bootstrap(path?:string): boolean {
    path = path ?? process.cwd() ?? "./";
    const config = tsConfigPaths.loadConfig(path);

    if (config.resultType !== "success") {
        throw new Error(`Unable to load config: ${path}`);
    }
    tsConfigPaths.register({
        baseUrl: config.absoluteBaseUrl,
        paths: config.paths,
    });
    return true
}

let done: boolean = false;
(() => {
    done || (bootstrap() && (done = true))
})()
```

and in my index simply `import "bootstrap";` which preps the paths. 

I should note. I **only** have issues using tsconfig paths @alias when the module contains function exports. If the module ONLY contains interfaces being exported then it's perfectly fine. The moment I get anywhere near exporting a function I get the module not found error.

`lib/types.ts`
```
export interface LibType {
    readonly val:string,
}
```

`lib/funcs.ts`
```
import './types`;
export function doFunc(v:LibType): LibType {
  return {val:'return'} as LibType;
}
```

Importing `@lib/types` is fine, importing `@lib/funcs` fails. Merging together fails...

**Comment by maelp** [comment]
Still hitting this issue, any idea? It should be much easier to import other typescript module from Pulumi code

**Comment by maelp** [comment]
Ah actually it works by adding in Pulumi.yaml

```
options:
  nodeargs: "--loader ts-node/esm --no-warnings"
```

**Comment by maelp** [comment]
BTW when I use the standard Pulumi ts config and I add a `src` dir with eg

```
src/
   index.ts
   mymodule.ts
   othermodule.ts
```

and that I try `import * from "./mymodule"` it fails, and I need to do `import * from "./mymodule.js"` which is a bit annoying, any way to fix this?

**Comment by matthew-gladman-oua** [comment]
This still seems to be a problem in 2026, for me the simplest option was just to switch to the `bun` runtime instead of trying to fix this issue. 

I'm sure it's workaroundable with `ts-node` or `tsx` (I did briefly try `tsx` but it didn't really work)

Example:

```yaml
# Pulumi.yaml
runtime:
  name: bun
  options:
    packagemanager: npm
```

```jsonc
# package.json

{
  "devDependencies": {
    "@types/node": "24.12.2",
    "bun": "1.3.11",
    "typescript": "6.0.2"
  }
}
```
